### PR TITLE
Dynamically allocate arrays using `MAX_SCREENWIDTH`/`HEIGHT`

### PR DESCRIPTION
**Pull request by MrAlaux**
Alright, let's get some automated builds and test performance.

### CODE REVIEW

**Comment by liPillON** [comment]
Hi, finally had the time to try the build resulting from this pr, here are the result.
For comparison, I've included a bunch of data from previous tests.

It seems that this PR has brought Nugget's performances in line with Woof 11.2
The only exceptions being: 
- demonastery (better than nugg-master, worst than woof-11.2)
- alien vendetta (better than both nugg-master and woof-11.2)
- pln.wad (same as nugg-master, worst than woof-11.2)

Hopefully when the optimizations from Woof will be merged in Nugget, the performance boost will compensate the costs of making 800p and 1600p available.

I'm still curious about testing the current PR with `MAX_HIRES 2`

<table>
<tr>
 <th>pwad             <th>woof 11.2 <th>woof master <th>nugg master <th>nugg pr45 
<tr>                                                                              
 <td>pln              <td>63        <td>106         <td>33          <td>34
<tr>                                                                              
 <td>eviternity       <td>315       <td>338         <td>290         <td>312
<tr>                                                                              
 <td>alien vendetta   <td>575       <td>609         <td>550         <td>596
<tr>                                                                              
 <td>saturnine chapel <td>260       <td>268         <td>245         <td>258
<tr>                                                                              
 <td>grove            <td>249       <td>258         <td>238         <td>250
<tr>                                                                              
 <td>demonastery      <td>471       <td>479         <td>415         <td>431
</table>

**Comment by MrAlaux** [comment]
Hmm... I take it these tests were performed by warping to some maps and standing still, right? Any chance you could try some timedemos for each build?

**Comment by liPillON** [comment]
With the exception of pln.wad, all tests were timedemos (three runs each, average fps).
This is the same methodology I've used for the results posted in #44 


win11 x64
intel i7 1260p
16gb ddr4 dual channel
iris xe graphics
1920x1200 display
-noautoload -nosound
clean-from-scratch cfg (only changes: vsync off, widescreen auto, demo progeessbar on)
between each run I waited for cpu/gpu cool off, no thermal throttling should have happened

**Comment by MrAlaux** [comment]
I see, thanks.

@ceski-1 thoughts?

**Comment by ceski-1** [comment]
Looks like it's time to review `visplane_t` again.

**Comment by liPillON** [comment]
for the record, the optimizations I've mentioned were introduced in Woof with these PRs:
https://github.com/fabiangreffrath/woof/pull/1108
https://github.com/fabiangreffrath/woof/pull/1110
https://github.com/fabiangreffrath/woof/pull/1111

File: src/r_bsp.c [hunk 1]
```diff
@@ -124,7 +124,7 @@ static void R_ClipWallSegment(int first, int last, boolean solid)
 
 void R_ClearClipSegs (void)
 {
-  memset(solidcol, 0, MAX_SCREENWIDTH);
+  memset(solidcol, 0, SCREENWIDTH << hires); // [Nugget] Dynamic arrays
```
Comment (fabiangreffrath) [inline src/r_bsp.c:127]: You should be able to get away with `viewwidth` whenever you use `SCREENWIDTH << hires` in this code change.

File: src/r_draw.c [hunk 2]
```diff
@@ -823,6 +821,12 @@ void R_InitBuffer(int width, int height)
 
   // Column offset. For windows.
 
+  // [Nugget] Dynamic arrays
+  if (first_allocation)
```
Comment (fabiangreffrath) [inline src/r_draw.c]: Shouldn't be necessary, just make sure arrays point to `NULL` when uninitialized. `realloc(NULL, ...)` is supposed to be the same as `malloc(...)`.

File: src/r_draw.c [hunk 3]
```diff
@@ -47,8 +44,8 @@ int  scaledviewheight;        // killough 11/98
 int  viewheight;
 int  viewwindowx;
 int  viewwindowy; 
-byte *ylookup[MAXHEIGHT]; 
-int  columnofs[MAXWIDTH]; 
+byte **ylookup; // [Nugget] Dynamic arrays
```
Comment (fabiangreffrath) [inline src/r_draw.c]: Global variables should always be implicitly initialized to `NULL`, but I would explicitly add it here anyway.

File: src/r_main.c [hunk 4]
```diff
@@ -687,12 +687,31 @@ void R_ExecuteSetViewSize (void)
...
+  }
+  #undef width
+  
   R_InitData();
   R_SetViewSize(screenblocks);
   R_InitPlanes();
   R_InitLightTables();
   R_InitSkyMap();
   R_InitTranslationTables();
+  R_InitSprites(sprnames); // [Nugget] Moved from P_Init()
```
Comment (fabiangreffrath) [inline src/r_main.c:703]: Why did you move this around at all?

File: src/z_zone.h [hunk 5]
```diff
@@ -40,6 +40,7 @@
 
 typedef enum {
   PU_STATIC,
+  PU_VIDEO, // [Nugget] Dynamic arrays
```
Comment (fabiangreffrath) [inline src/z_zone.h]: I don't get it. Why is this necessary?

**Comment by liPillON** [comment]
BTW I re-run all timedemos because while testing the latest PR I noticed some inconsistencies..
Here are the results: I believe they are more trustworthy since this time I used the same methodology for all tests (the previous table was simple an accumulation of old results).

All timedemos were ran 5 times (letting the cpu cool down in between runs). 
Best and Worst results were excluded.
I then calculated the average of the remaining 3 results.

For PLN.WAD I simply warped into the level and stood still in one corner of the map, looking at the opposite corner.

Noteworthy considerations:
- nugg 1.14 performing better than woof 11.2
- sensible performance gain in woof master (artifacts run 798: various optimizations)
- sensible performance hit in nugget master (run69: supporting rendering resolutions up to 1600p)
- even stronger performance hit with this PR (pr45, run67: dynamic array allocations)

In particular, the last point confirms what @MrAlaux was anticipating here https://github.com/MrAlaux/Nugget-Doom/issues/44#issuecomment-1598110422

Sorry for the misleading previous results...



|TIMEDEMOS            |woof 11.2 |woof run798 |nugg 1.14 |nugg run69 |nugg pr45 |
|---------------------|----------|------------|----------|-----------|----------|
|alien vendetta MAP20 |605       |643         |637       |612        |582       |
|demonastery          |458       |508         |498       |460        |448       |
|eviternity MAP26     |337       |358         |347       |324        |312       |
|grove                |256       |266         |264       |253        |248       |
|planisfere 2         |363       |379         |370       |343        |331       |
|saturnine chapel     |272       |279         |275       |263        |258       |
|sunlust MAP30        |275       |285         |278       |258        |256       |
                                              
                                              
                                              
|VISPLANE TEST MAP    |woof 11.2 |woof run798 |nugg 1.14 |nugg run69 |nugg pr45 |
|---------------------|----------|------------|----------|-----------|----------|
|PLN.WAD              |60        |100         |59        |35         |35        |


DEMOS:
https://dsdarchive.com/files/demos/av/28588/av20-1357.zip
https://dsdarchive.com/files/demos/demonastery/48438/demonastery-1326.zip
https://dsdarchive.com/files/demos/eviternity/44536/evit26-658.zip
https://dsdarchive.com/files/demos/grove/12807/grov-724.zip
https://dsdarchive.com/files/demos/planisf2/48969/planisf2os910.zip
https://dsdarchive.com/files/demos/satchap/20510/satchap-706.zip
https://dsdarchive.com/files/demos/sunlust/57817/sl30m2321.zip

File: src/r_main.c (hunk 4 above)
Comment (MrAlaux) [reply src/r_main.c:703]: I moved it because of `*negonearray`. I _could_ have just initialized it (and every other array) in `R_Init()`, but I think it's cleaner if they're initialized in the init function of the module where they're declared.

File: src/z_zone.h (hunk 5 above)
Comment (MrAlaux) [reply src/z_zone.h]: This implementation is loosely based on code from Eternity, which added a `PU_VALLOC` tag to be used for the memory allocated for these arrays. I figured I should add a new tag too, but I must admit I've no idea of what exactly it is for. What other tag do you think I should use?

Comment (fabiangreffrath) [reply src/z_zone.h]: Whenever you use `Z_Malloc()` in a way where you would usually use plain `malloc()` (i.e. without any magic involved to automatically free resources e.g. when finishing a map) that'd be `PU_STATIC`.

**Comment by MrAlaux** [comment]
Thanks for this, @liPillON. 

It seems I was right about this PR's effects, then. If these changes are not improving performance in their current state, I don't think I'll be able to figure out how to make them do so. I'll try tweaking the allocation process as @fabiangreffrath suggested, but as it stands, that as far as I can go.

Nugget 11.4 performing better than Woof 11.2 was a complete surprise, I wonder how that happened.

The performance hit of allowing higher resolutions is certainly unwanted, but I guess we can live with it considering the unexpected performance gain over Woof, and also the fact that further down the road, Woof's latest optimizations will be merged. I might make a branch merging those optimizations early so you can test it.

**Comment by liPillON** [comment]
Sure! When the time comes, feel free to @ me in the discussion for issue 44 

While we're here: what about benchmarking a `MAX_HIRES 2` branch?

File: src/r_main.c (hunk 4 above)
Comment (fabiangreffrath) [reply src/r_main.c:703]: But `negonearray` is really just an array full of `-1`. I don't see much value in having this one allocated dynamically. I would just leave it static and save the hacks.

Comment (MrAlaux) [reply src/r_main.c:703]: Right, makes sense.

**Comment by MrAlaux** [comment]
I changed some things based on most of Fabian's suggestions, but to no avail it seems; the only change that seems to have increased performance was making `negonearray` static again, hinting at the fact that making arrays dynamic causes a performance loss. I don't think these last changes are worth benchmarking.

I'm pretty sure that I'm the one doing something wrong here, and I'm certain I won't be able to figure out what exactly. As it stands, I can't do anything else with this branch, but I guess I'll leave the PR open for a while in case anyone wants to chime in.
